# Part 1: Forward Mode Automatic Differentiation

Forward mode AD can simply be implemented by defining a class to represent [dual numbers](https://en.wikipedia.org/wiki/Dual_number) which hold the value and its derivative. The following skeleton defines a dual number and implements multiplication. 

__Tasks:__

- Addition (`__add__`) is incomplete - can you finish it? 
- Can you also implement division (`__truediv__`), subtraction (`__sub__`) and power (`__pow__`)?

In [0]:
import math

class DualNumber:
    def __init__(self, value, dvalue):
        self.value = value
        self.dvalue = dvalue

    def __str__(self):
        return str(self.value) + " + " + str(self.dvalue) + "ε"

    def __mul__(self, other):
        return DualNumber(self.value * other.value,
            self.dvalue * other.value + other.dvalue * self.value)
    
    def __add__(self, other):
        #TODO: finish me
        # YOUR CODE HERE
        return DualNumber(self.value + other.value,
                         self.dvalue + other.dvalue)
        #raise NotImplementedError()
    
    # TODO: add missing methods
    # YOUR CODE HERE
    def __truediv__(self, other):
        return DualNumber(self.value / other.value,
                         (self.dvalue * other.value - self.value * other.dvalue) / (other.value * other.value))
    
    def __sub__(self, other):
        return DualNumber(self.value - other.value,
                         self.dvalue - other.dvalue)
     
    def __pow__(self,other):
        return DualNumber(math.pow(self.value,other.value), math.pow(self.value,other.value) * (other.dvalue * math.log(self.value) + self.dvalue * other.value / self.value))
        
        
    #raise NotImplementedError()

In [0]:
# Tests

DualNumber(1,0) + DualNumber(1,0) / DualNumber(1,0) - DualNumber(1,0)**DualNumber(1,0)


## Implementing math functions

We also need to implement some core math functions. Here's the sine function for a dual number:

In [0]:
def sin(x):
    return DualNumber(math.sin(x.value), math.cos(x.value)*x.dvalue)

__Task:__ can you implement the _cosine_ (`cos`), _tangent_ (`tan`), and _exponential_ (`exp`) functions in the code block below?

In [0]:
# TODO: implement additional math functions on dual numbers

def cos(x):
    # YOUR CODE HERE
    return DualNumber(math.cos(x.value), -math.sin(x.value) * x.dvalue)
    #raise NotImplementedError()

def tan(x):
    # YOUR CODE HERE
    return DualNumber(math.tan(x.value), x.dvalue / (math.cos(x.value) * math.cos(x.value)))
    #raise NotImplementedError()

def exp(x):
    # YOUR CODE HERE
    # Using Euler's formula
    return DualNumber(math.exp(x.value) * math.cos(x.dvalue), math.exp(x.value) * math.sin(x.dvalue))
    #raise NotImplementedError()

In [0]:
# Tests
assert cos(DualNumber(0,0)).value == 1
assert tan(DualNumber(0,0)).value == 0
assert exp(DualNumber(0,0)).value == 1


## Time to try it out

We're now in a position to try our implementation.

__Task:__ 

- Try running the following code to compute the value of the function $z=x\cdot y+sin(x)$ given $x=0.5$ and $y=4.2$, together with the derivative $\partial z/\partial x$ at that point. 

In [0]:
# YOUR CODE HERE
x = DualNumber(0.5,0)
y = DualNumber(4.2,0)
dNumber_dx = DualNumber(0.5,1) # value as x and dvalue as derivare of 1
fx = dNumber_dx * y + sin(dNumber_dx)
print("For the given x and y, z is", fx.value, ", and dzdx is", fx.dvalue)
#print(4.2 + math.cos(0.5))
#raise NotImplementedError()

For the given x and y, z is 2.579425538604203 , and dzdx is 5.077582561890373


__Task__: Differentiate the above function with respect to $x$ and write the symbolic derivatives in the following box. Verify the result computed above is correct by plugging-in the values into your symbolic gradient expression.

$\partial z/\partial x$ = y + cos(x) = 4.2 + cos(0.5) = 5.07758

__Task:__ Now use the code block below to compute the derivative $\partial z/\partial y$ of the above expression (at the same point $x=0.5, y=4.2$ as above) and store the derivative in the variable `dzdy` (just the derivative, not the Dual Number). Verify by hand that the result is correct.

In [0]:
# YOUR CODE HERE
dNumber_dy = DualNumber(4.2,1)
fy = dNumber_dy * x
dzdy = fy.dvalue
#raise NotImplementedError()

print('dz/dy:', dzdy)

dz/dy: 0.5


In [0]:
#Tests
assert dzdy
assert type(dzdy) == float


__Task:__ Finally, use the code block below to experiment and test the other math functions and methods you created.

In [0]:
# YOUR CODE HERE
# Try y = x**4.3 + 3x, given x  = 3.2
index = DualNumber(4.3,0)
dNumber_x = DualNumber(3.2,1)
y = pow(dNumber_x,index) + DualNumber(3,0) * dNumber_x
print("dydx is :", y.dvalue)
print(4.3 * (3.2**3.3) + 3)
#raise NotImplementedError()

dydx is : 202.73923584481346
202.7392358448135
